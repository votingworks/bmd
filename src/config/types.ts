// Generic
export interface Dictionary<T> {
  [key: string]: T | undefined
}

// Events
export type InputEvent = React.FormEvent<EventTarget>
export type ButtonEvent = React.MouseEvent<HTMLButtonElement>
export type ButtonEventFunction = (event: ButtonEvent) => void

// Candidates
export interface Candidate {
  readonly id: string
  readonly name: string
  readonly partyId?: string
  isWriteIn?: boolean
}
export type OptionalCandidate = Candidate | undefined

// Contests
export type ContestTypes = 'candidate' | 'yesno'
export interface Contest {
  readonly id: string
  readonly districtId: string
  readonly partyId?: string
  readonly section: string
  readonly title: string
  readonly type: ContestTypes
}
export interface CandidateContest extends Contest {
  readonly type: 'candidate'
  readonly seats: number
  readonly candidates: Candidate[]
  readonly allowWriteIns: boolean
}
export interface YesNoContest extends Contest {
  readonly type: 'yesno'
  readonly description: string
  readonly shortTitle: string
}
export type Contests = (CandidateContest | YesNoContest)[]

// Election
export interface BMDConfig {
  readonly requireActivation?: boolean
  readonly showHelpPage?: boolean
  readonly showSettingsPage?: boolean
}
export interface ElectionDefaults {
  readonly bmdConfig: BMDConfig
}
export interface BallotStyle {
  readonly id: string
  readonly precincts: string[]
  readonly districts: string[]
  readonly partyId?: string
}
export interface Party {
  readonly id: string
  readonly name: string
  readonly abbrev: string
}
export type Parties = Party[]
export interface Precinct {
  readonly id: string
  readonly name: string
}
export interface District {
  readonly id: string
  readonly name: string
}
export interface County {
  readonly id: string
  readonly name: string
}

export interface Election {
  readonly ballotStyles: BallotStyle[]
  readonly county: County
  readonly parties: Parties
  readonly precincts: Precinct[]
  readonly districts: District[]
  readonly contests: Contests
  readonly date: string
  readonly seal?: string
  readonly sealURL?: string
  readonly state: string
  readonly title: string
  readonly bmdConfig?: BMDConfig
}
export type OptionalElection = Election | undefined

export interface ActivationData {
  ballotStyle: BallotStyle
  precinct: Precinct
}

// Votes
export type CandidateVote = Candidate[]
export type YesNoVote = 'yes' | 'no'
export type OptionalYesNoVote = YesNoVote | undefined
export type Vote = CandidateVote | YesNoVote
export type OptionalVote = Vote | undefined
export type VotesDict = Dictionary<Vote>

// Ballot
export type UpdateVoteFunction = (contestId: string, vote: OptionalVote) => void
export type MarkVoterCardUsedFunction = (props: {
  ballotPrinted: boolean
}) => Promise<boolean>
export interface BallotContextInterface {
  activateBallot: (activationData: ActivationData) => void
  ballotStyleId: string
  contests: Contests
  readonly election: Election | undefined
  incrementBallotsPrintedCount: () => void
  isLiveMode: boolean
  markVoterCardUsed: MarkVoterCardUsedFunction
  precinctId: string
  resetBallot: (path?: string) => void
  setUserSettings: (partial: PartialUserSettings) => void
  updateVote: UpdateVoteFunction
  userSettings: UserSettings
  votes: VotesDict
}

// Smart Card Content
export type CardDataTypes = 'voter' | 'pollworker' | 'clerk'
export interface CardData {
  readonly t: CardDataTypes
}
export interface VoterCardData extends CardData {
  readonly t: 'voter'
  readonly bs: string
  readonly pr: string
  readonly uz?: number
  readonly bp?: number
}
export interface PollworkerCardData extends CardData {
  readonly t: 'pollworker'
  readonly h: string
}
export interface ClerkCardData extends CardData {
  readonly t: 'clerk'
  readonly h: string
}

// User Interface
export type ScrollDirections = 'up' | 'down'
export interface ScrollShadows {
  showBottomShadow: boolean
  showTopShadow: boolean
}
export interface Scrollable {
  isScrollable: boolean
}

export type TextSizeSetting = 0 | 1 | 2 | 3

export interface UserSettings {
  textSize: TextSizeSetting
}
export type PartialUserSettings = Partial<UserSettings>

export default {}
