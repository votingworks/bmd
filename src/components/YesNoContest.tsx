import React from 'react'
import styled from 'styled-components'

import {
  ButtonEvent,
  InputEvent,
  OptionalYesNoVote,
  ScrollDirections,
  UpdateVoteFunction,
  YesNoContest as YesNoContestInterface,
  YesNoVote,
} from '../config/types'

import BallotContext from '../contexts/ballotContext'

import GLOBALS from '../config/globals'
import Button from './Button'
import Main from './Main'
import Modal from './Modal'
import Prose from './Prose'
import { TextWithLineBreaks } from './Text'

const tabletMinWidth = 720
const votes = GLOBALS.YES_NO_VOTES

const ContentHeader = styled.div`
  width: 100%;
  max-width: 35rem;
  margin: 0px auto;
  padding: 1rem 0.75rem 0.5rem;
  @media (min-width: ${tabletMinWidth}px) {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
`
const ContestSection = styled.div`
  font-size: 0.85rem;
  font-weight: 600;
  text-transform: uppercase;
`
const ContestDescription = styled.div`
  padding: 0 0.25rem;
  @media (min-width: ${tabletMinWidth}px) {
    padding-left: 4rem;
  }
`
const ContestFooter = styled.div`
  width: 100%;
  max-width: 35rem;
  margin: 0 auto;
  padding: 1rem 0.5rem;
  @media (min-width: ${tabletMinWidth}px) {
    padding-right: 1rem;
    padding-left: 1rem;
  }
`
const VariableContentContainer = styled.div<{
  showBottomShadow?: boolean
  showTopShadow?: boolean
}>`
  display: flex;
  flex: 1;
  position: relative;
  overflow: auto;
  &:before,
  &:after {
    content: '';
    z-index: 1;
    transition: opacity 0.25s ease;
    position: absolute;
    height: 0.25rem;
    width: 100%;
  }
  &:before {
    top: 0;
    opacity: ${({ showTopShadow }) =>
      showTopShadow ? /* istanbul ignore next: Tested by Cypress */ 1 : 0};
    background: linear-gradient(
      to bottom,
      rgb(177, 186, 190) 0%,
      transparent 100%
    );
  }
  &:after {
    bottom: 0;
    opacity: ${({ showBottomShadow }) =>
      showBottomShadow ? /* istanbul ignore next: Tested by Cypress */ 1 : 0};
    background: linear-gradient(
      to bottom,
      transparent 0%,
      rgb(177, 186, 190) 100%
    );
  }
`
const ScrollControls = styled.div`
  display: none;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  flex-direction: column;
  justify-content: space-between;
  width: 100%;
  max-width: 35rem;
  padding: 0.5rem 0.75rem 0.5rem 0;
  padding-left: calc(100% - 7rem);
  pointer-events: none;
  & > * {
    pointer-events: auto;
  }
  html[data-useragent*='Windows'] & {
    margin-left: -17px; /* Windows Chrome scrollbar width */
  }
  @media (min-width: ${tabletMinWidth}px) {
    display: flex;
  }
  @media (min-width: 840px) {
    left: 50%;
    margin-left: -420px;
    padding-left: calc(840px - 7rem);
    html[data-useragent*='Windows'] & {
      margin-left: calc(-420px + -17px); /* Windows Chrome scrollbar width */
    }
  }
`
const ScrollContainer = styled.div`
  flex: 1;
  overflow: auto;
`
const ScrollableContentWrapper = styled.div<{ isScrollable: boolean }>`
  width: 100%;
  max-width: 35rem;
  margin: 0 auto;
  padding: 0.5rem 0.5rem 1rem;
  @media (min-width: ${tabletMinWidth}px) {
    padding-right: ${({ isScrollable }) =>
      isScrollable
        ? /* istanbul ignore next: Tested by Cypress */ '8rem'
        : '1rem'};
    padding-left: 1rem;
  }
`
const ChoicesGrid = styled.div`
  display: grid;
  grid-auto-rows: minmax(auto, 1fr);
  grid-gap: 0.75rem;
`
const Choice = styled('label')<{ isSelected: boolean }>`
  cursor: pointer;
  position: relative;
  display: grid;
  min-height: 2.5rem;
  align-items: center;
  border-radius: 0.125rem;
  background: ${({ isSelected }) => (isSelected ? '#028099' : 'white')};
  color: ${({ isSelected }) => (isSelected ? 'white' : undefined)};
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0, 0, 0, 0.14),
    0 0.1875rem 0.0625rem -0.125rem rgba(0, 0, 0, 0.12),
    0 0.0625rem 0.3125rem 0 rgba(0, 0, 0, 0.2);
  transition: background 0.25s, color 0.25s;
  button& {
    text-align: left;
  }
  :focus-within {
    outline: -webkit-focus-ring-color auto 5px;
  }
  :before {
    content: '${({ isSelected }) => (isSelected ? 'âœ“' : '')}';
    position: absolute;
    left: 0;
    top: 0;
    bottom: 0;
    background: white;
    border-right: 1px solid;
    border-color: ${({ isSelected }) => (isSelected ? '#028099' : 'lightgrey')};
    width: 3rem;
    text-align: center;
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 2rem;
    border-radius: 0.125rem 0 0 0.125rem;
    color: #028099;
  }
  & > div {
    word-break: break-word;
    padding: 0.5rem 0.5rem 0.5rem 4rem;
    @media (min-width: 480px) {
      padding: 1rem 1rem 1rem inherit;
    }
  }
`
const ChoiceInput = styled.input.attrs({
  type: 'checkbox',
})`
  margin-right: 0.5rem;
`
interface Props {
  contest: YesNoContestInterface
  vote: OptionalYesNoVote
  updateVote: UpdateVoteFunction
}

interface State {
  isScrollable: boolean
  isScrollAtBottom: boolean
  isScrollAtTop: boolean
  overvoteSelection: OptionalYesNoVote
}

const initialState = {
  isScrollable: false,
  isScrollAtBottom: true,
  isScrollAtTop: true,
  overvoteSelection: undefined,
}

export default class YesNoContest extends React.Component<Props> {
  public static contextType = BallotContext
  public state: State = initialState
  private scrollContainer = React.createRef<HTMLDivElement>()

  public componentDidMount() {
    this.updateContestChoicesScrollStates()
    window.addEventListener('resize', this.updateContestChoicesScrollStates)
  }

  public componentDidUpdate(prevProps: Props) {
    if (this.props.vote !== prevProps.vote) {
      this.updateContestChoicesScrollStates()
    }
  }

  public handleDisabledClick = () => {
    // maybe we'll do more when a disabled item is clicked, for now nothing.
  }

  public handleUpdateSelection = (event: InputEvent) => {
    const target = event.target as HTMLInputElement
    const newVote = target.value as YesNoVote
    const { vote } = this.props
    if (vote === newVote) {
      this.props.updateVote(this.props.contest.id, undefined)
    } else {
      this.props.updateVote(this.props.contest.id, newVote)
    }
  }

  public handleChangeVoteAlert = (overvoteSelection: YesNoVote) => {
    this.setState({ overvoteSelection })
  }

  public updateContestChoicesScrollStates = () => {
    const target = this.scrollContainer.current
    /* istanbul ignore next */
    if (!target) {
      return
    }
    const isTabletMinWidth = target.offsetWidth >= tabletMinWidth
    const targetMinHeight =
      GLOBALS.FONT_SIZES[this.context.userSettings.textSize] * 8 // magic number: room for buttons + spacing
    const windowsScrollTopOffsetMagicNumber = 1 // Windows Chrome is often 1px when using scroll buttons.
    const windowsScrollTop = Math.ceil(target.scrollTop) // Windows Chrome scrolls to sub-pixel values.
    this.setState({
      isScrollable:
        isTabletMinWidth &&
        /* istanbul ignore next: Tested by Cypress */
        target.scrollHeight > target.offsetHeight &&
        /* istanbul ignore next: Tested by Cypress */
        target.offsetHeight > targetMinHeight,
      isScrollAtBottom:
        windowsScrollTop +
          target.offsetHeight +
          windowsScrollTopOffsetMagicNumber >= // Windows Chrome "gte" check.
        target.scrollHeight,
      isScrollAtTop: target.scrollTop === 0,
    })
  }

  public scrollContestChoices = /* istanbul ignore next: Tested by Cypress */ (
    event: ButtonEvent
  ) => {
    const direction = (event.target as HTMLElement).dataset
      .direction as ScrollDirections
    const scrollContainer = this.scrollContainer.current!
    const currentScrollTop = scrollContainer.scrollTop
    const offsetHeight = scrollContainer.offsetHeight
    const scrollHeight = scrollContainer.scrollHeight
    const idealScrollDistance = Math.round(offsetHeight * 0.75)
    const maxScrollableDownDistance =
      scrollHeight - offsetHeight - currentScrollTop
    const maxScrollTop =
      direction === 'down'
        ? currentScrollTop + maxScrollableDownDistance
        : currentScrollTop
    const idealScrollTop =
      direction === 'down'
        ? currentScrollTop + idealScrollDistance
        : currentScrollTop - idealScrollDistance
    const top = idealScrollTop > maxScrollTop ? maxScrollTop : idealScrollTop
    scrollContainer.scrollTo({
      behavior: 'smooth',
      left: 0,
      top,
    })
  }

  public closeOvervoteAlert = () => {
    this.setState({
      overvoteSelection: initialState.overvoteSelection,
    })
  }

  public render() {
    const { contest, vote } = this.props
    const { overvoteSelection } = this.state
    const { isScrollable, isScrollAtBottom, isScrollAtTop } = this.state
    return (
      <React.Fragment>
        <Main noOverflow noPadding>
          <ContentHeader id="contest-header">
            <Prose>
              <h1 aria-label={`${contest.section}, ${contest.title}.`}>
                <ContestSection>{contest.section}</ContestSection>
                {contest.title}
              </h1>
              <p>
                <strong>Vote Yes or No.</strong>
              </p>
            </Prose>
          </ContentHeader>
          <VariableContentContainer
            showTopShadow={!isScrollAtTop}
            showBottomShadow={!isScrollAtBottom}
          >
            <ScrollContainer
              ref={this.scrollContainer}
              onScroll={this.updateContestChoicesScrollStates}
            >
              <ScrollableContentWrapper isScrollable={isScrollable}>
                <ContestDescription>
                  <Prose>
                    <TextWithLineBreaks text={contest.description} />
                  </Prose>
                </ContestDescription>
              </ScrollableContentWrapper>
            </ScrollContainer>
            {isScrollable /* istanbul ignore next: Tested by Cypress */ && (
              <ScrollControls aria-hidden="true">
                <Button
                  data-direction="up"
                  disabled={isScrollAtTop}
                  onClick={this.scrollContestChoices}
                  tabIndex={-1}
                >
                  â†‘ See More
                </Button>
                <Button
                  data-direction="down"
                  disabled={isScrollAtBottom}
                  onClick={this.scrollContestChoices}
                  tabIndex={-1}
                >
                  â†“ See More
                </Button>
              </ScrollControls>
            )}
          </VariableContentContainer>
          <ContestFooter>
            <ChoicesGrid role="group" aria-labelledby="contest-header">
              {['Yes', 'No'].map(answer => {
                const answerLowerCase = answer.toLowerCase()
                const isChecked = vote === answerLowerCase
                const isDisabled = !isChecked && !!vote
                const handleDisabledClick = () => {
                  if (isDisabled) {
                    this.handleChangeVoteAlert(
                      answer.toLowerCase() as YesNoVote
                    )
                  }
                }
                return (
                  <Choice
                    key={answer}
                    htmlFor={answerLowerCase}
                    isSelected={isChecked}
                    onClick={handleDisabledClick}
                  >
                    <ChoiceInput
                      id={answerLowerCase}
                      name={contest.id}
                      value={answerLowerCase}
                      onChange={
                        isDisabled
                          ? this.handleDisabledClick
                          : this.handleUpdateSelection
                      }
                      checked={isChecked}
                      className="visually-hidden"
                    />
                    <Prose>
                      <p>{answer}</p>
                    </Prose>
                  </Choice>
                )
              })}
            </ChoicesGrid>
          </ContestFooter>
        </Main>
        <Modal
          isOpen={!!overvoteSelection}
          content={
            <Prose>
              {overvoteSelection && (
                <p>
                  Do you want to change your vote to{' '}
                  <strong>{votes[overvoteSelection]}</strong>? To change your
                  vote, first unselect your vote for{' '}
                  <strong>
                    {
                      {
                        no: votes.yes,
                        yes: votes.no,
                      }[overvoteSelection]
                    }
                  </strong>
                  .
                </p>
              )}
            </Prose>
          }
          actions={
            <Button primary autoFocus onClick={this.closeOvervoteAlert}>
              Okay
            </Button>
          }
        />
      </React.Fragment>
    )
  }
}
