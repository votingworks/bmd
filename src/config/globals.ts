export default {
  CARD_POLLING_INTERVAL: 200,
  CHECK_ICON: 'âœ“',
  FONT_SIZES: [18, 24, 28, 32],
  TEXT_SIZE: 1,
  YES_NO_VOTES: {
    no: 'No',
    yes: 'Yes',
  },
}
