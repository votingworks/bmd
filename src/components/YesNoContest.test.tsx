import React from 'react'
import { fireEvent, render } from 'react-testing-library'

import { YesNoContest as YesNoContestInterface } from '../config/types'

import YesNoContest from './YesNoContest'

const contest: YesNoContestInterface = {
  description: 'description',
  districtId: 'district-id',
  id: 'contest-id',
  section: 'County',
  shortTitle: 'Prop 1',
  title: 'Prop 1: Garbage Collection Program',
  type: 'yesno',
}

describe('supports yes/no contest', () => {
  it(`allows voting for both yes and no`, () => {
    const updateVote = jest.fn()
    const { container, getByText } = render(
      <YesNoContest
        contest={contest}
        vote={undefined}
        updateVote={updateVote}
      />
    )
    expect(container).toMatchSnapshot()

    fireEvent.click(getByText('Yes').closest('label')!)
    expect(updateVote).toHaveBeenCalledTimes(1)

    fireEvent.click(getByText('No').closest('label')!)
    expect(updateVote).toHaveBeenCalledTimes(2)
  })
  it(`displays warning when attempting to change vote`, () => {
    const updateVote = jest.fn()
    const { container, getByText } = render(
      <YesNoContest contest={contest} vote={'yes'} updateVote={updateVote} />
    )
    expect(container).toMatchSnapshot()
    fireEvent.click(getByText('No').closest('label')!)
    expect(
      getByText(
        (_, node) =>
          node.textContent ===
          'Do you want to change your vote to No? To change your vote, first unselect your vote for Yes.'
      )
    )
    // expect(getByTestId('modal-content').textContent).toEqual(
    //   'Do you want to change your vote to No? To change your vote, first unselect your vote for Yes.'
    // )
    fireEvent.click(getByText('Okay'))
  })
})
