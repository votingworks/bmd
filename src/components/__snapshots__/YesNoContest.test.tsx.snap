// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports yes/no contest allows voting for both yes and no 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  width: 100%;
  max-width: 66ch;
  line-height: 1.2;
}

.c2 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c2 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c2 h3,
.c2 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
  font-size: 1rem;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p {
  margin-top: -0.75rem;
}

.c2 h3 + p {
  margin-top: -1rem;
}

.c2 > :first-child {
  margin-top: 0;
}

.c2 > :last-child {
  margin-bottom: 0;
}

.c12 {
  word-break: break-word;
}

.c1 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 1rem 0.75rem 0.5rem;
}

.c3 {
  text-transform: uppercase;
  font-size: 0.85rem;
  font-weight: 600;
}

.c7 {
  padding: 0 0.25rem;
}

.c8 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 1rem 0.5rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: auto;
}

.c4::before,
.c4::after {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 0.25rem;
  content: '';
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c4::before {
  top: 0;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c4::after {
  bottom: 0;
  opacity: 0;
  background: linear-gradient( to bottom, transparent 0%, rgb(177,186,190) 100% );
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c6 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 0.5rem 0.5rem 1rem;
}

.c9 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
}

.c10 {
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #FFFFFF;
  cursor: pointer;
  min-height: 2.5rem;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c10 {
  text-align: left;
}

.c10:focus-within {
  outline: rgb(77,144,254) dashed 0.25rem;
}

.c10::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: rgb(211,211,211);
  border-radius: 0.125rem 0 0 0.125rem;
  background: #FFFFFF;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  content: '';
}

.c10 > div {
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c11 {
  margin-right: 0.5rem;
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

@media (min-width:480px) {
  .c2 {
    line-height: 1.3;
  }
}

@media (min-width:720px) {
  .c1 {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
}

@media (min-width:720px) {
  .c7 {
    padding-left: 4rem;
  }
}

@media (min-width:720px) {
  .c8 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:720px) {
  .c6 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:480px) {
  .c10 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
      id="contest-header"
    >
      <div
        class="c2"
        id="audiofocus"
      >
        <h1
          aria-label="Prop 1: Garbage Collection Program."
        >
          <div
            class="c3"
          >
            County
          </div>
          Prop 1: Garbage Collection Program
        </h1>
        <p
          aria-label="Vote Yes or No. Use the down arrow to select your preference. Use the right arrow to move to the next contest."
        >
          <strong>
            Vote Yes or No.
          </strong>
        </p>
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <div
              class="c2"
            >
              <p>
                description
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c8"
    >
      <div
        class="c9"
      >
        <label
          class="c10"
          for="yes"
        >
          <input
            class="visually-hidden c11"
            id="yes"
            name="contest-id"
            role="option"
            type="checkbox"
            value="yes"
          />
          <div
            class="c2"
          >
            <p
              aria-label="Yes on Prop 1"
              class="c12"
            >
              Yes
            </p>
          </div>
        </label>
        <label
          class="c10"
          for="no"
        >
          <input
            class="visually-hidden c11"
            id="no"
            name="contest-id"
            role="option"
            type="checkbox"
            value="no"
          />
          <div
            class="c2"
          >
            <p
              aria-label="No on Prop 1"
              class="c12"
            >
              No
            </p>
          </div>
        </label>
      </div>
    </div>
  </main>
</div>
`;

exports[`supports yes/no contest displays warning when attempting to change vote 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  width: 100%;
  max-width: 66ch;
  line-height: 1.2;
}

.c2 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c2 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c2 h3,
.c2 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
  font-size: 1rem;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p {
  margin-top: -0.75rem;
}

.c2 h3 + p {
  margin-top: -1rem;
}

.c2 > :first-child {
  margin-top: 0;
}

.c2 > :last-child {
  margin-bottom: 0;
}

.c12 {
  word-break: break-word;
}

.c1 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 1rem 0.75rem 0.5rem;
}

.c3 {
  text-transform: uppercase;
  font-size: 0.85rem;
  font-weight: 600;
}

.c7 {
  padding: 0 0.25rem;
}

.c8 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 1rem 0.5rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: auto;
}

.c4::before,
.c4::after {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 0.25rem;
  content: '';
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c4::before {
  top: 0;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c4::after {
  bottom: 0;
  opacity: 0;
  background: linear-gradient( to bottom, transparent 0%, rgb(177,186,190) 100% );
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c6 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 0.5rem 0.5rem 1rem;
}

.c9 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
}

.c13 {
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #FFFFFF;
  cursor: pointer;
  min-height: 2.5rem;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c13 {
  text-align: left;
}

.c13:focus-within {
  outline: rgb(77,144,254) dashed 0.25rem;
}

.c13::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: rgb(211,211,211);
  border-radius: 0.125rem 0 0 0.125rem;
  background: #FFFFFF;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  content: '';
}

.c13 > div {
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c10 {
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #028099;
  cursor: pointer;
  min-height: 2.5rem;
  color: #FFFFFF;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c10 {
  text-align: left;
}

.c10:focus-within {
  outline: rgb(77,144,254) dashed 0.25rem;
}

.c10::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: #028099;
  border-radius: 0.125rem 0 0 0.125rem;
  background: #FFFFFF;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  content: '✓';
}

.c10 > div {
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c11 {
  margin-right: 0.5rem;
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

@media (min-width:480px) {
  .c2 {
    line-height: 1.3;
  }
}

@media (min-width:720px) {
  .c1 {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
}

@media (min-width:720px) {
  .c7 {
    padding-left: 4rem;
  }
}

@media (min-width:720px) {
  .c8 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:720px) {
  .c6 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:480px) {
  .c13 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

@media (min-width:480px) {
  .c10 > div {
    padding: 1rem 1rem 1rem inherit;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
      id="contest-header"
    >
      <div
        class="c2"
        id="audiofocus"
      >
        <h1
          aria-label="Prop 1: Garbage Collection Program."
        >
          <div
            class="c3"
          >
            County
          </div>
          Prop 1: Garbage Collection Program
        </h1>
        <p
          aria-label="Vote Yes or No. Use the down arrow to select your preference. Use the right arrow to move to the next contest."
        >
          <strong>
            Vote Yes or No.
          </strong>
        </p>
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <div
              class="c2"
            >
              <p>
                description
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c8"
    >
      <div
        class="c9"
      >
        <label
          class="c10"
          for="yes"
        >
          <input
            checked=""
            class="visually-hidden c11"
            id="yes"
            name="contest-id"
            role="option"
            type="checkbox"
            value="yes"
          />
          <div
            class="c2"
          >
            <p
              aria-label="Yes on Prop 1"
              class="c12"
            >
              Yes
            </p>
          </div>
        </label>
        <label
          class="c13"
          for="no"
        >
          <input
            class="visually-hidden c11"
            id="no"
            name="contest-id"
            role="option"
            type="checkbox"
            value="no"
          />
          <div
            class="c2"
          >
            <p
              aria-label="No on Prop 1"
              class="c12"
            >
              No
            </p>
          </div>
        </label>
      </div>
    </div>
  </main>
</div>
`;
